flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -270,9 +270,22 @@ private void restartTasksWithDelay(final FailureHandlingResult failureHandlingRe
         delayExecutor.schedule(
                 () ->
                         FutureUtils.assertNoException(
-                                cancelFuture.thenRunAsync(
-                                        restartTasks(executionVertexVersions, globalRecovery),
-                                        getMainThreadExecutor())),
+                                cancelFuture
+                                        .thenRunAsync(
+                                                restartTasks(
+                                                        executionVertexVersions, globalRecovery),
+                                                getMainThreadExecutor())
+                                        .thenRunAsync(
+                                                () ->
+                                                        archiveExceptions(
+                                                                failureHandlingResult.getError(),
+                                                                executionVertexVersions.stream()
+                                                                        .map(
+                                                                                ExecutionVertexVersion
+                                                                                        ::getExecutionVertexId)
+                                                                        .collect(
+                                                                                Collectors
+                                                                                        .toList())))),
```

Comment:
I think this change goes in a good direction. But what we need to do before resetting the `Executions` in `restartTasks` is to collect all failure from the set of reset/restarted `Executions` and group them together with the `rootCause` into a failure information object.

Reply:
I think in some way you are already doing it with the `exceptionCache`. It might be easier to read the `Execution.failureCause` for this.

Reply:
There is also a list of prior `Executions` in `ExecutionVertex.priorExecutions`.

Reply:
Thanks for the pointer to `priorExecutions`. I missed it out completely.
You're saying that instead of caching the `Throwable`s in `SchedulerBase::executionCache` I should use `Execution::priorExecution`?! That makes sense as we avoid saving the same information twice. But I would have to change the `priorCause` to `ErrorInfo` (instead of `Throwable`) to also collect the time when the `Throwable` was caught. Am I right?

Reply:
No, actually I think a better way would be to ask the `Executions` which are reset whether they have failed or not and if they have to record the `Execution.failureCause`. For the timestamp you can check whether taking the state transition timestamp to the `FAILED` state is good enough.